### PR TITLE
Add internal domain model

### PR DESCRIPTION
**Pull request by yurishkuro**

### CODE REVIEW

File: model/keyvalue.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	if one.Key != two.Key {
+		return one.Key < two.Key
+	}
+	if one.VType != two.VType {
+		return one.VType < two.VType
+	}
+	switch one.VType {
+	case StringType:
+		return one.VStr < two.VStr
+	case BoolType:
```
Comment (black-adder) [inline model/keyvalue.go]: nit: can combine BoolType and Int64Type

File: model/keyvalue.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+// Binary returns the blob ([]byte) value stored in this KeyValue or nil if it stores a different type.
+func (kv *KeyValue) Binary() []byte {
+	if kv.VType == BinaryType {
+		return kv.VBlob
+	}
+	return nil
+}
+
+// IsLess compares two KeyValue objects. The order is based first on the keys, then on type, and finally no the value.
```
Comment (black-adder) [inline model/keyvalue.go]: s/no/on

File: model/keyvalue.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	switch one.VType {
+	case StringType:
+		return one.VStr < two.VStr
+	case BoolType:
+		return one.VNum < two.VNum
+	case Int64Type:
+		return one.VNum < two.VNum
+	case Float64Type:
+		return one.Float64() < two.Float64()
+	case BinaryType:
```
Comment (black-adder) [inline model/keyvalue.go]: can't we automatically assume that if the lengths are different we can exit early with `return l1 < l2`?

File: model/keyvalue.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	return KeyValue{Key: key, VType: BinaryType, VBlob: value}
+}
+
+// Bool returns the Boolean value stored in this KeyValue or false if it stores a different type.
+func (kv *KeyValue) Bool() bool {
+	if kv.VType == BoolType {
+		return kv.VNum == 1
+	}
+	return false
+}
```
Comment (badiib) [inline model/keyvalue.go]: what's the difference between a boolean and something that's not a boolean? we need to return error.

File: model/process.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+// Process describes an instance of an application or service that emits tracing data.
+type Process struct {
+	ServiceName string `json:"serviceName"`
+	Tags        []Tag  `json:"tags,omitempty"`
+}
+
+// NewProcess creates a new Process for given serviceName and tags.
+// The tags are sorted in place and kept in the the same array/slice.
+func NewProcess(serviceName string, tags []Tag) *Process {
+	SortTags(tags)
```
Comment (badiib) [inline model/process.go]: why sort the tags?

File: model/keyvalue.go (hunk 4 above)
Comment (yurishkuro) [inline model/keyvalue.go]: The assumption is that you check for type before reading a typed value. If you do that, you never get an error situation, so I would rather not introduce errors into getters.

File: model/keyvalue.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+import "math"
+
+// ValueType describes the type of value contained in a KeyValue struct
+type ValueType int
+
+const (
+	// StringType indicates the value is a unicode string
+	StringType ValueType = iota
+	// BoolType indicates the value is a Boolean encoded as int64 number 0 or 1
+	BoolType
```
Comment (yurishkuro) [inline model/keyvalue.go]: That would mean we need to do heap allocation every type we want to store a bool or a number. A binary blob is in the "never happens" category anyway, 99% of tags are strings, which would be already pre-allocated elsewhere, so in the current form we have no heap allocations when creating a tag.

**Comment by yurishkuro** [comment]
> You are assuming clients of this model will check the type before asking for the type.

The only way to make this safe is to introduce a visitor pattern, and I really don't want to do that. The most frequent use of this data will be either translating it into something else (like the UI model), or checking if a particular tag exists. The latter can be done with `==` against pre-defined constants

**Pull request merged**